woocommerce/woocommerce-android
# Hack Week: Switch in-app reviews to google's new Reviews API

This PR closes #2700 by implementing [Google's new In-App Reviews API](https://developer.android.com/guide/playcore/in-app-review?hl=en) which allows us to capture a play store review without leaving the app. 

### The old flow

When clicking "I like it" a dialog would appear asking the user if they'd like to leave us a review. If the user chose to leave a review they'd be routed out of the app and over to the Google Play app. If the user had already left a review there wouldn't be anything for them to do, they'd just go back to the app.

<img src="https://user-images.githubusercontent.com/5810477/93522791-02c4f580-f900-11ea-92b2-49e848abe2f7.gif" width="320"/>

### The new flow

Now when the user clicks "I like it" we will submit a request to start the in-app review process. **_It may or may not actually happen_**. This is because the new API enforces **_quotas_** - if the user has already left a review, or if they've already hit the maximum times we can ask them to review - the reviews flow will not start at all. We have no control over this part of the process and will not be notified what happened or if the user left a review. Here's how the google docs explain this part of the process:
> To provide a great user experience, Google Play enforces a quota on how often a user can be shown the review dialog. Because of this, calling a launchReviewFlow method might not always display a dialog. For example, you should not have a call-to-action option (such as a button) to trigger a review as a user might have already hit their quota and the flow won’t be shown, presenting a broken experience to the user.

**Here's how it works if the user has hit max quota:**
1. User clicks "I like it"
2. The Feedback card disappears. 
3. We submit a request to start the review flow but because the user has already hit quota, the flow never starts. 

![no-review](https://user-images.githubusercontent.com/5810477/93523552-36544f80-f901-11ea-856b-96262373183f.gif)


**Here's how it works if the user hasn't hit max quota:**
1. User clicks "I like it"
2. The Feedback card disappears. 
3. We submit a request to start the review flow and the in-app review ratings card is displayed. 
4. User fills out the review and submits. 

![review](https://user-images.githubusercontent.com/5810477/93523565-3ce2c700-f901-11ea-8184-7d7311e71333.gif)

### What's not included

I left the `AppRatingDialog` intact for now. If we decide that we like this new reviews flow I can retrofit that class to use the new API instead of routing the user to the Google Play Store. 

### To Test
Testing this one is super tricky because first the Feedback card has to be visible, which it will only do if you've had the app for installed for more than 90 days, and then you would've had to have not hit quota for reviews which is also really difficult. The only way to get around this limitation was to create a build that _always_ shows the Feedback card and load it up on the Play store for "internal app sharing". This special bucket bypasses the rules of quotas so you'll always see the full review flow, you just won't be able to actually submit a review since it's a sandbox. Just ping me with your test google email and I'll add you as a tester and send you a link. 

Update release notes:
